### PR TITLE
Jolie Import System with Symbols and Resolvers

### PR DESCRIPTION
**Pull request by kicito**
The Implementation for Jolie's import system, Currently supports type definitions and interfaces

Syntax:

> importStmt ::= 'from' '.'* <modules> 'import' '*' | <symbol> ['as' <symbol>]

The implementation is based on https://app.gitbook.com/@jolielang/s/developer-guide/interpreter-documentation/parsing-and-loading

Additional changes in existed codebase:
- SemanticVerifier is no longer resolve TypeDefinitionLink. The work is now responsible by GlobalSymbolResolver
- OOITBuilder: New algorithm for resolveTypeLinks, resolves from the paired typedefintion rather than the local types table
- OLParser:
  - resturcture of the parseLoop function, in order to prepare for an introduction of servicenode. 
  - Symbols ast nodes (type, interfaces) is now assigned as a empty named definition. Which to be resolved by GlobalSymbolReferenceResolver

example 
```
// server.ol
include "console.iol"

from .iface import i
from .type import testType

interface internalIface{
    OneWay: notice(testType)
}

inputPort ip{
    Interfaces: i,internalIface
    Protocol: sodep
    Location: "socket://localhost:3000"
}

execution{concurrent}

main{
    [twice(req)(res){
        res = req*2
    }]
    [notice(req)]{
        println@Console(req.sub1)()
    }
}
```

client.ol
```
include "console.iol"

from .iface import i
from .type import testType

outputPort op{
    Interfaces: i
    OneWay: notice(testType)
    Protocol: sodep
    Location: "socket://localhost:3000"
}

main{
    twice@op(2)(res)
    println@Console(res)()
    notice@op(void{.sub1="someStr"})
}
```

```
//iface.ol
interface i {
    RequestResponse: twice(int)(int)
}

//type.ol


type testType: void{
    sub1: string
}

// private symbol, cannot be imported by external module
private type privateTestType: void{
    sub2: string
}
```

### CODE REVIEW

**Comment by fmontesi** [comment]
Looks very good. Do you have any way of testing it under Windows?

**Comment by bmaschio** [comment]
@fmontesi  I can do it

**Comment by bmaschio** [comment]
I launched the mvn install phace this my result

comparisons.ol...          passed.
correlation_set.ol...      passed.
deepCopy.ol...             passed.
defined.ol...              passed.
Warning: Nashorn engine is planned to be removed from a future JDK release
embedded_js.ol...          passed.
expressions.ol...          passed.
faultHandling.ol...        passed.
foreach_array.ol...        passed.
foreach_nodes.ol...        passed.
inline_trees.ol...         passed.
instanceof.ol...           passed.
interfaceExtender.ol...    passed.
internal_service.ol...     passed.
jap.ol...                  passed.
newline_sequences.ol...    passed.
operations.ol...           passed.
redirection.ol...          passed.
self_call.ol...            passed.
service_block.ol...        passed.
spawn.ol...                passed.
sr_timeout.ol...           passed.
types.ol...                passed.
with.ol...                 passed.
converter.ol...            failed. strings invalid length/size (1)
giu 15, 2020 2:53:26 PM converter.ol
INFO: [converter.ol] Thrown unhandled fault: TestFailed
Content (if any) = strings invalid length/size (1) : string

files.ol...                passed.
inspector.ol...            passed.
json_utils.ol...           passed.
message_digest.ol...       passed.
metajolie.ol...            passed.
metarender.ol...           passed.
reflection.ol...           passed.
time.ol...                 passed.
uri_templates.ol...        passed.
xml_utils.ol...            passed.
zip_utils.ol...            passed.
auto.ol...                 passed.
http_cookies.ol...         passed.
http_features.ol...        passed.
http_get.ol...             passed.
http_post.ol...            passed.
include_http.ol...         passed.
local.ol...                passed.
server.ol...               passed.

**Comment by fmontesi** [comment]
@bmaschio 
Mmh, that doesn't look right, it's not running the test primitives/import.ol

Maybe you're on the wrong branch? @kicito , does the test run on your machine?

**Comment by bmaschio** [comment]
@fmontesi 

C:\Progetti\Github\kicito\jolie>git config --get remote.origin.url
https://github.com/kicito/jolie.git

C:\Progetti\Github\kicito\jolie>git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

C:\Progetti\Github\kicito\jolie>

**Comment by fmontesi** [comment]
The URL is right, but I think it's in branch `jolie-import-system`, not in branch `master`.

**Comment by bmaschio** [comment]
@fmontesi 

import.ol...               jolie.runtime.embedding.EmbeddedServiceLoadingException: jolie.runtime.embedding.EmbeddedServiceLoadingException: jolie.InterpreterException: jolie.lang.parse.module.ModuleException: C:\Progetti\Github\kicito\jolie\test\primitives\import.ol:8: error: jolie.lang.parse.module.exceptions.ModuleNotFoundException: Module "main" not found from lookup path [twice\twice\main.ol, twice\main in lib\twice.jap]
        at jolie.runtime.embedding.JolieServiceLoader.load(JolieServiceLoader.java:66)
        at joliex.lang.RuntimeService.loadEmbeddedService(RuntimeService.java:249)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at jolie.runtime.JavaService.requestResponseCallable(JavaService.java:239)
        at jolie.runtime.JavaService.callOperation(JavaService.java:322)
        at jolie.runtime.embedding.JavaCommChannel.lambda$sendImpl$0(JavaCommChannel.java:85)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: jolie.runtime.embedding.EmbeddedServiceLoadingException: jolie.InterpreterException: jolie.lang.parse.module.ModuleException: C:\Progetti\Github\kicito\jolie\test\primitives\import.ol:8: error: jolie.lang.parse.module.exceptions.ModuleNotFoundException: Module "main" not found from lookup path [twice\twice\main.ol, twice\main in lib\twice.jap]
        at jolie.runtime.embedding.JolieServiceLoader.load(JolieServiceLoader.java:63)
        ... 13 more
Caused by: jolie.InterpreterException: jolie.lang.parse.module.ModuleException: C:\Progetti\Github\kicito\jolie\test\primitives\import.ol:8: error: jolie.lang.parse.module.exceptions.ModuleNotFoundException: Module "main" not found from lookup path [twice\twice\main.ol, twice\main in lib\twice.jap]
        at jolie.Interpreter.buildOOIT(Interpreter.java:1335)
        at jolie.Interpreter.init(Interpreter.java:1091)
        at jolie.Interpreter.access$1000(Interpreter.java:110)
        at jolie.Interpreter$StarterThread.run(Interpreter.java:1220)
Caused by: jolie.lang.parse.module.ModuleException: C:\Progetti\Github\kicito\jolie\test\primitives\import.ol:8: error: jolie.lang.parse.module.exceptions.ModuleNotFoundException: Module "main" not found from lookup path [twice\twice\main.ol, twice\main in lib\twice.jap]
        at jolie.lang.parse.module.ModuleCrawler.crawlModule(ModuleCrawler.java:93)
        at jolie.lang.parse.module.ModuleCrawler.crawl(ModuleCrawler.java:102)
        at jolie.Interpreter.buildOOIT(Interpreter.java:1283)
        ... 3 more
Caused by: jolie.lang.parse.module.exceptions.ModuleNotFoundException: Module "main" not found from lookup path [twice\twice\main.ol, twice\main in lib\twice.jap]
        at jolie.lang.parse.module.AbsolutePathFinder.find(Finder.java:197)
        at jolie.lang.parse.module.ModuleCrawler.findModule(ModuleCrawler.java:81)
        at jolie.lang.parse.module.ModuleCrawler.crawlModule(ModuleCrawler.java:89)
        ... 5 more

inline_trees.ol...         passed.

**Comment by fmontesi** [comment]
:o) Windows...
@kicito could you have a look? Maybe the code paths we have in CommandLineParser and URIUtils could be helpful.

File: jolie/src/main/java/jolie/CommandLineParser.java [hunk 1]
```diff
@@ -70,6 +70,7 @@
 	private final int connectionsCache;
 	private final CorrelationEngine.Type correlationAlgorithmType;
 	private final String[] includePaths;
```
Comment (fmontesi) [inline jolie/src/main/java/jolie/CommandLineParser.java:72]: We have recently modified CommandLineParser to produce an InterpreterParameters object, so this pull request needs to be rebased/merged and modified to account for that. (Should be a small modification.)

File: libjolie/src/main/java/jolie/lang/parse/module/Finder.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		}
+		packagesTokenStartIndex = i;
+		return basePath;
+	}
+
+	@Override
+	public Source find() throws ModuleNotFoundException {
+		Path basePath = resolveDotPrefix();
+		try {
+			Source moduleFile = super.locateModule( basePath );
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Finder.java:161]: This code is very hard to follow, because you rely on the side-effect that resolveDotPrefix changes what packagesToken will return inside of super.locateModule.

I'd remove the packagesToken method entirely, and have that resolveDotPrefix returns you a composite object containing both the basePath and the array of "packagesToken" (maybe consider calling it pathParts instead? It took me a while to figure out what packagesToken is).

Then, have `locateModule` take the "packagesToken" thing as a second parameter every time.

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	public ModuleCrawler( Path workingDirectory, String[] packagesPath, ModuleParser parser )
+		throws FileNotFoundException {
+		this( new FinderCreator( workingDirectory, packagesPath ), parser );
+	}
+
+	private Source findModule( URI parentURI, String[] importTargetStrings )
+		throws ModuleNotFoundException {
+		Finder finder = finderCreator.getFinderForTarget( parentURI, importTargetStrings );
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java:80]: This smells of over-engineering (YAGNI). You create a finder, call only one of its methods with no parameters, and then immediately discard it.
Consider renaming FinderCreator to Finder and replace `getFinderForTarget` with `find`.

Then `Finder.find` can implement the logic you already have, but without having all these subclasses for `Finder`.

File: libjolie/src/main/java/jolie/lang/parse/module/Finder.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+
+	@Override
+	public Source find() throws ModuleNotFoundException {
+		/**
+		 * 1. Try to resolve P directly from WDIR. 2. Check if FIRST.jap is in WDIR/lib. If so, resolve REST
+		 * inside of this jap. 3. Try to resolve P from the list of packages directories.
+		 */
+		ModuleNotFoundException err = new ModuleNotFoundException( this.moduleName() );
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Finder.java:197]: It's weird that you create an exception you might not need. Consider using a list of messages in this method instead.

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	public ModuleRecord parse( Source module ) throws ParserException, IOException, ModuleException {
+		String[] additionalPath;
+		if( module.includePath().isPresent() ) {
+			additionalPath = new String[] { module.includePath().get() };
+		} else {
+			additionalPath = new String[ 0 ];
+		}
+		String[] inculdePaths =
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java:85]: includePaths

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		program = OLParseTreeOptimizer.optimize( program );
+		SymbolTable st = SymbolTableGenerator.generate( program );
+		return new ModuleRecord( scanner.source(), program, st );
+	}
+
+	public ModuleRecord parse( URI uri, String[] additionalIncludePaths, boolean joinPaths )
+		throws ParserException, IOException, ModuleException {
+		URL url = uri.toURL();
+		InputStream stream = url.openStream();
+		String[] inculdePaths = includePaths;
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java:129]: includePaths (also below)

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleRecord.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	/**
+	 * @return the program
+	 */
+	public Program program() {
+		return program;
+	}
+
+	public void setSymbolTable( SymbolTable symbolTable ) {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleRecord.java:56]: Never used. Remove?

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+	/**
+	 * pointer to an AST node
+	 */
+	private OLSyntaxNode node;
+
+	/**
+	 * privacy of this symbol
+	 */
+	final private Privacy privacy;
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java:58]: Prefer `private final` over `final private`.

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+				externalSymbol.setModuleSource( moduleSource );
+				modulesToCrawl.add( moduleSource );
+			} catch( ModuleNotFoundException e ) {
+				throw new ModuleException( externalSymbol.context(), e );
+			}
+		}
+		this.result.addModuleRecord( record );
+	}
+
+	public ModuleCrawlerResult crawl( ModuleRecord parentRecord )
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java:99]: Could be a static method (you always create the module crawler right before).

File: libjolie/src/main/java/jolie/lang/parse/module/GlobalSymbolReferenceResolver.java [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+			resolver.resolve( md.program() );
+		}
+	}
+
+	/**
+	 * Resolve symbols the is imported from external modules and resolve linked type's pointer
+	 * 
+	 * @throws ModuleException if the process is failed
+	 */
+	public void resolve() throws ModuleException {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/GlobalSymbolReferenceResolver.java:772]: Could be a static method (you always create the resolver right before).

Also, GlobalSymbolReferenceResolver could just be SymbolReferenceResolver? :-)

**Comment by kicito** [comment]
Thank you very much for your time @bmaschio @fmontesi. In the recent commit I have fix JAP issues and refactoring code based on your suggestion.

Apart from the typos and some unused methods, I changed ModuleCrawler, SymbolResolver to be called from via a static method and simplify the ModuleFinder (hopefully :) ) according to your comment. I was complicating stuff too much from studying module system in other language.

@thesave could you please have a look at the code?

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	public static void clearCache() {
+		ModuleCrawler.cache.clear();
+	}
+
+	private final Finder finder;
+	private final ModuleParser parser;
+
+	private ModuleCrawler( ModuleCrawlerComponent component ) {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java:85]: Erase the `ModuleCrawlerComponent` class and just bring the `Finder` and `ParserConfiguration` objects at this level.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	/**
+	 * resolve the wildcard symbol by replace with array of symbol rely in it's symbolTable
+	 * 
+	 * @param wildCardSymbol a wildcard symbol in symbol table to replace
+	 * @param symbolsFromWildcard array of symbols in parsed wildcardSymbol's module
+	 * @throws ModuleException when adding name duplicate name to the symbol
+	 */
+	public void replaceWildCardSymbol( SymbolWildCard wildCardSymbol,
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java:69]: rename to `resolveWildcardImport`

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+	 * create and add an external Symbol with it's module target to the table
+	 * 
+	 * @param context Context where symbol is declare
+	 * @param name Symbol name
+	 * @param moduleTargetStrings String array defining target module location defined at import
+	 *        statement
+	 * 
+	 * @throws ModuleException when adding name duplicate name to the symbol
+	 */
+	public void addSymbol( ParsingContext context, String name, List< String > moduleTargetStrings )
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java:107]: Make a class to represent import paths, called `ImportPath`. Otherwise we'll have to move these lists everywhere and they're not very informative. The name `moduleTargetStrings` is also not very informative, `importPath` will be clearer.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+	 */
+	public void addSymbol( String name, SymbolNode node ) throws DuplicateSymbolException {
+		if( isDuplicateSymbol( name ) ) {
+			throw new DuplicateSymbolException( name );
+		}
+		this.symbols.put( name, new SymbolInfoLocal( name, node ) );
+	}
+
+	/**
+	 * create and add an external Symbol with it's module target to the table
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java:98]: it's => its

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+	 * 
+	 * @param context Context where symbol is declare
+	 * @param name Symbol name
+	 * @param moduleTargetStrings String array defining target module location defined at import
+	 *        statement
+	 * @param moduleSymbol Name for binding result to local environment
+	 * 
+	 * @throws ModuleException when adding name duplicate name to the symbol
+	 */
+	public void addSymbol( ParsingContext context, String name, List< String > moduleTargetStrings,
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolTable.java:127]: rename to `addSymbolWithAlias`

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+	 * environment EXTERNAL means the symbol's AST node is declared in other execution environment
+	 */
+	public enum Scope {
+		LOCAL, EXTERNAL
+	}
+
+	/**
+	 * name of the symbol
+	 */
+	private String name;
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java:43]: final

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	public String name() {
+		return name;
+	}
+
+	/**
+	 * set a pointer to an AST node. the pointer should be only defined once
+	 */
+	public void setPointer( OLSyntaxNode pointer ) {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java:103]: rename to `resolve`.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+ * A class represent a Symbol defined within external execution environment. Create when consuming
+ * an import statement
+ */
+public class SymbolInfoExternal extends SymbolInfo {
+
+	private final List< String > moduleTargets;
+	private final String moduleSymbol;
+	private Source moduleSource;
+
+	public SymbolInfoExternal( ParsingContext context, String name, List< String > moduleTargets,
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java:36]: Do not represent module targets as `List`s. Use a dedicated `ImportPath` class.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java [hunk 19]
```diff
@@ -0,0 +1,782 @@
...
+			if( n.right() != null ) {
+				n.right().accept( this );
+			}
+		}
+
+		@Override
+		public void visit( ImportStatement n ) {}
+
+		private Optional< SymbolInfo > getSymbol( ParsingContext context, String name ) {
+			Optional< SymbolInfo > symbol = Helpers.firstNonNull( () -> {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:687]:
```
if( symbolTables.containsKey( currentURI ) && symbolTables.get( currentURI ).symbol( name ).isPresent() ) {
    return symbolTables.get( currentURI ).symbol( name );
} else if ( symbolTables.containsKey( context.source() ) && symbolTables.get( context.source() ).symbol( name ).isPresent() ) {
    return symbolTables.get( context.source() ).symbol( name );
}

return Optional.empty();
```

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java [hunk 20]
```diff
@@ -0,0 +1,782 @@
...
+	 */
+	public static void resolve( CrawlerResult moduleMap ) throws ModuleException {
+		SymbolReferenceResolver resolver = new SymbolReferenceResolver( moduleMap );
+		try {
+			resolver.resolveExternalSymbols();
+		} catch( SymbolNotFoundException | IllegalAccessSymbolException
+			| DuplicateSymbolException e ) {
+			throw new ModuleException( e );
+		}
+		resolver.resolveLinkedType();
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:780]: rename to `resolver.resolveLinkedTypes()`.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java [hunk 21]
```diff
@@ -0,0 +1,782 @@
...
+	 * 
+	 * @throws DuplicateSymbolException
+	 * @throws IllegalAccessSymbolException
+	 * @throws SymbolNotFoundException
+	 * 
+	 */
+	private void resolveExternalSymbols()
+		throws SymbolNotFoundException, IllegalAccessSymbolException, DuplicateSymbolException {
+		for( ModuleRecord md : moduleMap.values() ) {
+			for( SymbolInfoExternal localSymbol : md.externalSymbols() ) {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:737]: Calling an external symbol `localSymbol` is pure evil. :-D

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java [hunk 22]
```diff
@@ -0,0 +1,782 @@
...
+	 * perform lookup to external symbol's source of AST node
+	 * 
+	 * @throws SymbolNotFoundException
+	 */
+	private SymbolInfo symbolSourceLookup( SymbolInfoExternal symbolInfo )
+		throws SymbolNotFoundException {
+		ModuleRecord externalSourceRecord =
+			this.moduleMap.get( symbolInfo.moduleSource().get().source() );
+		Optional< SymbolInfo > externalSourceSymbol =
+			externalSourceRecord.symbol( symbolInfo.moduleSymbol() );
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:714]: rename `symbol` to `getSymbol`, since it's a lookup operation.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java [hunk 23]
```diff
@@ -0,0 +1,782 @@
...
+		if( !externalSourceSymbol.isPresent() ) {
+			throw new SymbolNotFoundException( symbolInfo.name(), symbolInfo.moduleTargets() );
+		}
+		if( externalSourceSymbol.get().scope() == Scope.LOCAL ) {
+			return externalSourceSymbol.get();
+		} else {
+			return symbolSourceLookup( (SymbolInfoExternal) externalSourceSymbol.get() );
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:721]: Can't this loop infinitely?
You need a parameter `Set< URI >` to remember the modules that you have already visited.

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	/**
+	 * set a destination source of the symbol
+	 */
+	public void setModuleSource( Source moduleSource ) {
+		this.moduleSource = moduleSource;
+	}
+
+	public List< String > moduleTargets() {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java:50]: rename to `importPath`

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import jolie.lang.parse.context.ParsingContext;
+
+/**
+ * A class represent a Symbol defined within external execution environment. Create when consuming
+ * an import statement
+ */
+public class SymbolInfoExternal extends SymbolInfo {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java:30]: rename to `ImportedSymbolInfo`

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoLocal.java [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package jolie.lang.parse.module;
+
+import jolie.lang.parse.ast.SymbolNode;
+
+/**
+ * A class represent Symbol define within local execution environment
+ */
+public class SymbolInfoLocal extends SymbolInfo {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoLocal.java:27]: Rename to `LocalSymbolInfo`

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+
+	/**
+	 * constructor for SymbolInfo, this constructor is used when it knows the ASTnode to point to
+	 * corresponding to this symbol
+	 * 
+	 * @param name Symbol name
+	 * @param scope scope of Symbol
+	 * @param node an ASTNode implementing SymbolNode
+	 */
+	public SymbolInfo( String name, Scope scope, SymbolNode node ) {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfo.java:73]: Make a single dumb constructor and have the subclasses decide which parameters to pass.

File: libjolie/src/main/java/jolie/lang/parse/module/Source.java [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+
+/**
+ * an Interface of Joile module Source
+ */
+public interface Source {
+
+	/**
+	 * @return URI location of the module
+	 */
+	URI source();
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Source.java:45]: rename to `uri()`

File: libjolie/src/main/java/jolie/lang/parse/module/Source.java [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import jolie.lang.Constants;
+
+/**
+ * an Interface of Joile module Source
+ */
+public interface Source {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Source.java:40]: rename to `ModuleSource`

File: libjolie/src/main/java/jolie/lang/parse/module/Source.java [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+
+	/**
+	 * @return an optional include path for parsing this module
+	 */
+	Optional< String > includePath();
+
+	/**
+	 * @return an InputStream of source
+	 */
+	Optional< InputStream > stream();
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Source.java:55]: rename to `openStream()`

File: libjolie/src/main/java/jolie/lang/parse/module/Source.java [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+
+/**
+ * Jolie Module in file (an .ol file)
+ */
+class FileSource implements Source {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Source.java:62]: Why not `Path`?

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+
+	public List< String > moduleTargets() {
+		return this.moduleTargets;
+	}
+
+	public Optional< Source > moduleSource() {
+		return Optional.of( this.moduleSource );
+	}
+
+	public String moduleSymbol() {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/SymbolInfoExternal.java:58]: rename to `originalSymbolName`.

File: libjolie/src/main/java/jolie/lang/parse/module/Finder.java [hunk 33]
```diff
@@ -0,0 +1,84 @@
...
+	 * @return Source Object which provides stream object for parser to use
+	 */
+	Source find( List< String > moduleTargetTokens, URI source ) throws ModuleNotFoundException;
+
+	/**
+	 * Returns an array of considering paths for resolving an absolute import of the finder.
+	 * 
+	 * @return an array of considering paths for resolving an absolute import
+	 */
+	Path[] packagePaths();
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/Finder.java:32]: Where is this used? Otherwise remove.

File: libjolie/src/main/java/jolie/lang/parse/module/FinderImpl.java [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+		File olTargetFile = Finder.olLookup( basePath, moduleName );
+		return new FileSource( olTargetFile );
+	}
+
+	/**
+	 * returns package paths this finder is considering
+	 * 
+	 * @return packagePaths
+	 */
+	public Path[] packagePaths() {
```
Comment (fmontesi) [inline libjolie/src/main/java/jolie/lang/parse/module/FinderImpl.java:148]: Unused?

File: libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java (hunk 21 above)
Comment (kicito) [reply libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java:737]: whopsss XD

File: jolie/src/main/java/jolie/Interpreter.java [hunk 35]
```diff
@@ -272,7 +277,7 @@ public CorrelationSet correlationInitializer() {
 	private final long persistentConnectionTimeout = 60 * 60 * 1000; // 1 hour
 	private final long awaitTerminationTimeout = 60 * 1000; // 1 minute
 
-
+	private Map< URI, SymbolTable > symbolTables;
```
[the remark this answers is not in the document]
Comment (kicito) [reply jolie/src/main/java/jolie/Interpreter.java:280]: This field is necessary in the future commit, when we enable the mechanism to embed a service node. The implementation was extended from embedded internal service where the loader initiate new interpreter with the current interpreter and parsed AST program. In service node version, the created interpreter accesses and passes parent's symboltable to semantic verification phase. Otherwise the verification phase in embed program would have no clue if the symbol has defined.

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleException.java [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+	public ModuleException( String message ) {
+		super( message );
+	}
+
+	public ModuleException( ParsingContext context, Throwable arg1 ) {
+		super( arg1 );
+		this.context = context;
+	}
+
+	public ModuleException( String arg0, Throwable arg1 ) {
```
[the remark this answers is not in the document]
Comment (kicito) [reply libjolie/src/main/java/jolie/lang/parse/module/ModuleException.java]: removed as they are unused

File: libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java [hunk 37]
```diff
@@ -0,0 +1,145 @@
...
+	public void putConstants( Map< String, Scanner.Token > constantsToPut ) {
+		constantsMap.putAll( constantsToPut );
+	}
+
+	public ModuleRecord parse( URI uri ) throws ParserException, IOException, ModuleException {
+		return parse( uri, this.includePaths, false );
+	}
+
+
+	public ModuleRecord parse( File file ) throws ParserException, IOException, ModuleException {
```
[the remark this answers is not in the document]
Comment (kicito) [reply libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java]: Agree, changed to Path.

[the remark this answers is not in the document]
Comment (bmaschio) [reply libjolie/src/main/java/jolie/lang/parse/module/ModuleParser.java]: @fmontesi  did we not use to have problems with path under windows?

File: jolie/src/main/java/jolie/Interpreter.java (hunk 35 above)
[the remark this answers is not in the document]
Comment (kicito) [reply jolie/src/main/java/jolie/Interpreter.java:276]: This field will be moved to InterpreterParameter
